### PR TITLE
dev: add msa transformer

### PR DESCRIPTION
**Pull request by PavelPodlesny**

### CODE REVIEW

File: alina/model/model.py [hunk 1]
```diff
@@ -3,14 +3,16 @@
...
+from .modules import ComplementaryLayer, EncoderLayer, MSATransformer, make_transformer_block
 
 
 class Model(nn.Module):
     def __init__(self,
                  vocab_size: int,
                  dim: int, 
+                 seq_len: int,
                  conv_layers: int,
+                 msa_transformers_order: int | List[int],
```
Comment (Arty40m) [inline alina/model/model.py]: Здесь теперь лучше назвать seq_encoders_order и struct_encoders_order (вместо encoders_order)
Так смысл будет отражаться хорошо

File: alina/model/model.py [hunk 2]
```diff
@@ -19,10 +21,13 @@ def __init__(self,
                 ):
         super().__init__()
         self.dim = dim
-        self.final_dot_norm = math.sqrt(dim)
+        self.final_dot_norm = math.sqrt(dim) # ???
```
Comment (Arty40m) [inline alina/model/model.py:24]: В чем вопрос
Это нормализация для финального скалярного произведения

File: alina/model/model.py [hunk 3]
```diff
@@ -19,10 +21,13 @@ def __init__(self,
...
         
         # Embedding
         self.embedding = nn.Embedding(vocab_size, dim)
+        self.nb_token = torch.nn.Parameter(torch.rand(1,1,dim), requires_grad=True) # NoBound token
```
Comment (Arty40m) [inline alina/model/model.py]: Это уже не токен, а эмбеддинг, вектор ведь
Поэтому nb_embed
И размерность лучше сделать просто dim, менять ее уже форварде через view, так суть веса чище выглядит

File: alina/model/model.py [hunk 4]
```diff
@@ -40,61 +45,89 @@ def __init__(self,
...
+        # b, msa -> b, 1, msa -> b, seq, msa -> b*seq, 1, 1, msa
+        msa_mask = msa_mask.unsqueeze(1).expand(-1, seq_len, -1).reshape(batch * seq_len, 1, 1, msa_len)
+        
+        for li in self.msa_transformers_order:
+            l = self.msa_transformers_block[li]
+            x = l(x, msa_mask, seq_mask)
+
+        x = x[:,0,:,:]   # batch, msa, seq, dim -> batch, seq, dim
+        # add nb token
+        nb_token = self.nb_token.expand(batch, 1, -1) # 1,1,dim -> b, 1, dim
```
Comment (Arty40m) [inline alina/model/model.py]: Сюда view перед экспандом

File: alina/model/modules.py [hunk 5]
```diff
@@ -2,7 +2,26 @@
 import torch
 import torch.nn as nn
 
+def make_transformer_block(transformer_layers_order, transformer_layer,
```
Comment (Arty40m) [inline alina/model/modules.py]: Аргументам добавить типы
Каждый на своей строке
Гиперпарамметры распаковать, их не так много чтобы паковать в тюпл, смысла нет

File: alina/model/modules.py [hunk 6]
```diff
@@ -45,52 +64,84 @@ def forward(self, x: torch.Tensor, struct_vec: torch.Tensor):
...
+        q = self.Q(q).view(-1, n, self.heads, self.depth)
+        k = self.K(k).view(-1, n, self.heads, self.depth)
+        v = self.V(v).view(-1, n, self.heads, self.depth)
+
+        q = q.permute(0, 2, 1, 3) # b, h, n, d
+        k = k.permute(0, 2, 1, 3) # b(0), h(1), n(2), d(3)
+        v = v.permute(0, 2, 1, 3) # b, h, n, d
+
+        if self.use_rope:
+            cos = self.rope_cos.to(q.dtype).to(q.device)
```
Comment (Arty40m) [inline alina/model/modules.py]: Перекидывать веса с цпу на гпу каждый форвард это очень плохая практика
Они должны лежать на девайсе модели
Можно сделать эти значения парамметром без градиента

File: alina/model/modules.py [hunk 7]
```diff
@@ -45,52 +64,84 @@ def forward(self, x: torch.Tensor, struct_vec: torch.Tensor):
...
+
+        self.use_rope = use_rope
+        self.seq_len = seq_len
         
         self.Q = nn.Linear(dim, dim)
         self.K = nn.Linear(dim, dim)
         self.V = nn.Linear(dim, dim)
         self.O = nn.Linear(dim, dim)
 
+        if self.use_rope:
```
Comment (Arty40m) [inline alina/model/modules.py]: ропе в таком случае лучше сделать торчевским модулем в парамметрами без градиента
В форвард модулю прилетят q k и он отдаст уже покрученные вектора
q, k = self.rope(q, k)

Если в этом блоке ропе использовать не нужно - на место модуля ропе поставить фейкованный ропе который ничего не делает:
self.rope = FakeRoPE()
q, k = self.rope(q, k) - эта инструкция не меняется в форварде аттеншена
Внутри фейка:
def forward(self, q, k): return q, k

В такой архитектуре не будет проверок ифов - ускорение выполнения графа

File: alina/model/modules.py [hunk 8]
```diff
@@ -45,52 +64,84 @@ def forward(self, x: torch.Tensor, struct_vec: torch.Tensor):
...
+        freq = torch.cat((freq, freq), dim=-1) # seq_len, depth
+        
+        self.rope_sin = freq.sin()[None,None,:,:] # 1,1,seq_len,depth
+        self.rope_cos = freq.cos()[None,None,:,:]
 
-        # batch, heads, seq, dim
-        q = q.permute(0, 2, 1, 3)
-        k = k.permute(0, 2, 3, 1)
-        v = v.permute(0, 2, 1, 3)
+    def swap_dims(self,x):
```
Comment (Arty40m) [inline alina/model/modules.py]: Это улетит в модуль ропе

File: alina/model/model.py [hunk 9]
```diff
@@ -3,14 +3,16 @@
...
 
-from .modules import ComplementaryLayer, EncoderLayer
+from .modules import ComplementaryLayer, EncoderLayer, MSATransformer, make_transformer_block
 
 
 class Model(nn.Module):
     def __init__(self,
                  vocab_size: int,
                  dim: int, 
+                 seq_len: int,
```
Comment (Arty40m) [inline alina/model/model.py]: Модель должна быть гибкой к длине сиквенса и необходимость каждый раз забивать тензор паддингом мне тоже не нравится. Мы уходим от Унета чтобы не зависить от конкретной длины.
Так что надо вернуться назад.
Поз энкодинг и ропе вычислять на лету по размерности тензора.
Да, это не так быстро, но не критично. По возможно нужно заготовить тензоры и вектора весов, которые не зависят от длины сиквенса, как парамметры и потом на лету вычислять остаток.

File: alina/dataset.py [hunk 10]
```diff
@@ -56,7 +56,7 @@ def to(self, device: str | torch.device):
                           lens=self.lens)
     
 
-class BaseInputPreprocessor:
```
Comment (Arty40m) [inline alina/dataset.py:59]: Чем так не угодили эти две буквы в слове input?)

File: alina/dataset.py [hunk 11]
```diff
@@ -56,7 +56,7 @@ def to(self, device: str | torch.device):
...
+class BaseInpPreprocessor:
     def __call__(self, na: nsk.NucleicAcid) -> torch.Tensor:
         inp_struct = torch.zeros(len(na)+1, dtype=torch.int32)
```
Comment (Arty40m) [inline alina/dataset.py]: Здесь длина + 1 для нб токена
Теперь его нет - надо убрать + 1

File: alina/dataset.py [hunk 12]
```diff
@@ -65,21 +65,45 @@ def __call__(self, na: nsk.NucleicAcid) -> torch.Tensor:
 
         return inp_struct
 
+class BaseOutPreprocessor:
```
Comment (Arty40m) [inline alina/dataset.py]: Лучше будет использовать один препроцессор и добавить операция инкремента индексов уже в моделе.
Я хочу чтобы мы вообще не задумывались об нб токене вне модели.
Будет как то так:
inp_struct = inp_struct + (inp_struct!=0).to(inp_struct.dtype) - то есть не инкрементим паддинг (это в форварде элины)

File: alina/dataset.py [hunk 13]
```diff
@@ -65,21 +65,45 @@ def __call__(self, na: nsk.NucleicAcid) -> torch.Tensor:
...
+    def __call__(self, na: nsk.NucleicAcid) -> torch.Tensor:
+        return torch.zeros(len(na)+1, dtype=torch.int32)
+        
+class EmptyOutPreprocessor:
+    def __call__(self, na: nsk.NucleicAcid) -> None: 
+        return torch.zeros(len(na), dtype=torch.int32)
 
+
+class AlinaDataset:
+    NT_MAP = {"A":2, "U":3, "G":4, "C":5, "N":6}
```
Comment (Arty40m) [inline alina/dataset.py]: 1 теперь используем
нб токена нет, он идет сразу в виде вектора весов

File: alina/dataset.py [hunk 14]
```diff
@@ -65,21 +65,45 @@ def __call__(self, na: nsk.NucleicAcid) -> torch.Tensor:
...
     def __init__(self,
                  nas: List[nsk.NucleicAcid],
-                 input_preprocessor: Callable[[nsk.NucleicAcid], torch.Tensor] = BaseInputPreprocessor(),
-                 cache: bool = True
+                 cache: bool = True,
+                 inp_struct_source: str | None = "struct",
+                 out_struct_source: str | None = None
                  ):
         
+        assert inp_struct_source in {None, "struct", *nas[0].meta}
```
Comment (Arty40m) [inline alina/dataset.py]: Здесь для меты лучше пройтись циклом и проверить во всех контейнерах, так 100% не упадем в ходе инференса

File: alina/dataset.py [hunk 15]
```diff
@@ -91,18 +115,35 @@ def __getitem__(self, n: int) -> AlinaDataPoint:
...
+        if self.out_struct_source not in (None, "struct"):
+            ss = nsk.NA(na.meta[self.out_struct_source])
+            out_struct = self.out_preprocessor(ss)
+        else: 
+            out_struct = self.out_preprocessor(na)
+
+        if "msa" in na.meta.keys():
+            msa = na.meta["msa"]
+            maxl = max([len(seq) for seq in msa])
+            seq_tensor = torch.zeros((self.MAX_MSA_LEN, maxl), dtype=torch.int32)
```
Comment (Arty40m) [inline alina/dataset.py]: Постоянный паддинг на 128?
Так не пойдет, слишком огромные тензоры
Паддинг должен выставляться в collate_fn

File: alina/dataset.py [hunk 16]
```diff
@@ -91,18 +115,35 @@ def __getitem__(self, n: int) -> AlinaDataPoint:
...
+            for i, seq in enumerate(msa):
+                t = torch.IntTensor([self.NT_MAP[nt] for nt in seq])
+                seq_tensor[i,:len(seq)] = t
+        else: 
+            maxl = len(na)
+            seq_tensor =torch.zeros((1, maxl), dtype=torch.int32)
+            seq_tensor[0] = torch.IntTensor([self.NT_MAP[nt] for nt in na.seq])
         
-        dp: AlinaDataPoint = AlinaDataPoint(seq=seq,
+        dp: AlinaDataPoint = AlinaDataPoint(seq=seq_tensor,
```
Comment (Arty40m) [inline alina/dataset.py:135]: Где в случае мса использование таргетного сиквенса?
Сначала идет na.seq, потом все что есть в мса

File: alina/dataset.py [hunk 17]
```diff
@@ -131,19 +172,21 @@ def __add__(self, other):
         
 
 def collate_fn(dps: List[AlinaDataPoint]) -> AlinaBatch:
+    MAX_MSA_LEN: int = 128
```
Comment (Arty40m) [inline alina/dataset.py]: Константа в двух местах это плохо
Лучше единую в шапку файла

File: alina/dataset.py [hunk 18]
```diff
@@ -131,19 +172,21 @@ def __add__(self, other):
...
     N: int = len(dps)
-    maxl: int = max([len(dp) for dp in dps])
+    #maxl: int = max([dp.len for dp in dps])
+    maxl = 256
```
Comment (Arty40m) [inline alina/dataset.py]: По комментам выше уже понятно - длина сиквенса - по самоу длинному представителю в батче. Все динамично, без лишнего паддинга

File: alina/model/model.py [hunk 19]
```diff
@@ -61,9 +63,9 @@ def __init__(self,
 
 
     @torch.compile
-    def pos_enc(self, seq: int, dim: int):
-        div  = 10000**(2 * (torch.arange(dim, dtype=torch.float32) // 2) / dim)
-        pos_enc = torch.arange(seq, dtype=torch.float32)
+    def pos_enc(self, seq: int, dim: int, device: torch.device):
+        div  = 10000**(2 * (torch.arange(dim, dtype=torch.float32, device=device) // 2) / dim)
+        pos_enc = torch.arange(seq, dtype=torch.float32, device=device)
```
Comment (Arty40m) [inline alina/model/model.py]: В torch.arange надо передать еще аргумент - стартовый индекс
Чтобы в метод можно было передать 1 - начинаем не с 0 а с 1

File: alina/model/model.py [hunk 20]
```diff
@@ -73,28 +75,50 @@ def pos_enc(self, seq: int, dim: int):
...
+        seq_mask = (seq==0).to(x.dtype).view(-1, 1, 1, seq_len) # b, msa, seq -> b+msa, 1, 1, seq
+        msa_mask = (seq.sum(dim=-1) == 0).to(x.dtype) # b, msa
+        # b, msa -> b, 1, msa -> b, seq, msa -> b*seq, 1, 1, msa
+        msa_mask = msa_mask.unsqueeze(1).expand(-1, seq_len, -1).reshape(batch * seq_len, 1, 1, msa_len)
+        
+        for li in self.seq_encoders_order:
+            l = self.seq_encoders_block[li]
+            x = l(x, msa_mask, seq_mask)
+
+        x = x[:,0,:,:]   # batch, msa, seq, dim -> batch, seq, dim
```
Comment (Arty40m) [inline alina/model/model.py:96]: Последовательность:
x = x[:,0,:,:] - берем тергетный сиквенс
x += self.pos_enc(... , start_idx=1) - прибавляем поз энкод и начинаем счет с 1 (0 для нб вектора)

struct_vec должен прийти с -1 для несвязанных уже из датасета иначе нельзя понять где несвязан а где связь с 0 нуклеотидом.
x = self.complementary_layer(x, struct_vec) - в struct_vec индексы начинаются с 0, -1 для свободных. Нужно сделать правки в графовый слой

Добавляем нб вектор к эмбедам у  которых уже есть инфа о вторичке
nb_embed = self.nb_embed.view(1,1,-1).expand(batch, 1, -1) # dim -> 1,1,dim -> b, 1, dim
x = torch.cat((nb_embed, x), dim=1) # b, seq+1, dim

Теперь маски
 seq = seq[:,0,:] # batch, msa, seq -> b, seq
 nb_pad = torch.ones((batch, 1), dtype=seq.dtype, device=seq.device) # b,1
 seq = torch.cat((nb_pad, seq), dim=1) # b, seq+1
 att_mask = (seq==0).to(x.dtype).view(seq.size(0), 1, 1, seq.size(1)) # b, seq -> b, 1, 1, seq

Дальше слои трансформера

File: alina/dataset.py [hunk 21]
```diff
@@ -91,18 +103,40 @@ def __getitem__(self, n: int) -> AlinaDataPoint:
...
+            out_struct = self.out_preprocessor(na)
+
+        if "msa" in na.meta.keys():
+            msa = na.meta["msa"]
+            msa_len = len(msa)
+            if msa[0]!=na.seq:
+                msa = [na.seq] + msa # prepend target seq to msa
+                msa_len+=1
+                
+            maxl = max([len(seq) for seq in msa])
```
Comment (Arty40m) [inline alina/dataset.py]: Все сиквенсы в мса должны быть одной длины
Здесь нужно наоборот сделать ассерт что они все одинаковы

File: alina/dataset.py [hunk 22]
```diff
@@ -91,18 +103,40 @@ def __getitem__(self, n: int) -> AlinaDataPoint:
...
+            seq_tensor = torch.zeros((msa_len, maxl), dtype=torch.int32)
+            for i, seq in enumerate(msa):
+                t = torch.IntTensor([self.NT_MAP[nt] for nt in seq])
+                seq_tensor[i,:len(seq)] = t
+        else:
```
Comment (Arty40m) [inline alina/dataset.py]: В такой логике идет дублирование кода
Лучше сделать как то так
Кидаем таргетный сиквенс в лист
Если есть мса в мете - идем циклом и добавляем их в лист, попутно добавляем эти сиквенсы в сэт и проверяем что нового нет в это сете, чтобы избежать дублей. Здесь же и проверка что длина одинаковая
Потом под этот лист уже делаем тензор и обрабатываем

File: alina/model/modules.py [hunk 23]
```diff
@@ -43,54 +64,108 @@ def forward(self, x: torch.Tensor, struct_vec: torch.Tensor):
...
-    def __init__(self, dim: int, heads: int):
+    def __init__(self, dim: int, heads: int, use_rope: bool):
         super().__init__()
         
         self.heads = heads
         self.dim = dim
         self.depth = dim//heads
         self.norm = math.sqrt(self.depth)
+
+        self.use_rope = use_rope
```
Comment (Arty40m) [inline alina/model/modules.py]: В целом можно этот параметр уже не сохранять
Нужный слой и так создает, а проверка не нужна
